neqo-transport/src/recovery.rs

```diff
@@ -654,57 +593,39 @@ impl LossRecovery {
             }
             qdebug!([self], "PTO timer fired for {}", space);
             pto_space = pto_space.or(Some(*space));
-            lost.extend(self.spaces[*space].pto_packets(PTO_COUNT).cloned());
+            lost.extend(self.spaces[*space].pto_packets(PTO_PACKET_COUNT).cloned());
         }
 
+        // This has to happen outside the loop. Increasing the PTO count here causes the
+        // pto_time to increase which might cause PTO for later packet number spaces to not fire.
         if let Some(space) = pto_space {
-            self.loss_recovery_state = LossRecoveryState::new(
-                LossRecoveryMode::PtoExpired {
-                    dgram_available: PTO_COUNT,
-                    min_pn_space: space,
-                },
-                Some(now),
-            );
-            self.pto_count += 1;
+            if let Some(st) = &mut self.pto_state {
+                st.pto(space);
+            } else {
+                self.pto_state = Some(PtoState::new(space));
+            }
         }
-        Some(lost)
     }
 
-    pub fn check_loss_detection_timeout(&mut self, now: Instant) -> Option<Vec<SentPacket>> {
-        qdebug!([self], "check_loss_timeouts");
-
-        if self.loss_recovery_state.mode() == LossRecoveryMode::None {
-            // LR not the active timer
-            return None;
-        }
+    pub fn timeout(&mut self, now: Instant) -> Vec<SentPacket> {
+        qtrace!([self], "timeout {:?}", now);
 
-        if self.callback_time() > Some(now) {
-            // LR timer, but hasn't expired.
-            return None;
+        let loss_delay = self.loss_delay();
+        let mut lost_packets = Vec::new();
+        for space in self.spaces.iter_mut() {
+            space.detect_lost_packets(now, loss_delay, &mut lost_packets);
         }
+        self.enable_timed_loss_detection = self.spaces.iter().any(|space| space.has_out_of_order());
+        self.maybe_pto(now, &mut lost_packets);
```

Comment:
Yeah, the PTO budget is now in recovery.rs (PtoState), which has a count (for the number of PTO timeouts) and `packets` for the number of packets that we can send on that PTO.

Reply:
For the first question, I wanted to double-check.  When something is declared lost using the usual loss detection function (`detect_lost_packets`), it will be added to the list, but not removed from the set of tracked packets.  It will therefore be considered again when the PTO "loss" event occurs.

That means that yes, a packet might appear in the final list twice.  But we have the existing guarantee that every sent item needs to be able to handle being declared lost 0 or more times, followed by maybe one time being acknowledged.  So this is fine from a robustness standpoint.

What this structure allows us to do is reduce the number of packets we declare "lost" on PTO if there are other packets we have already declared lost.  That seems like a reasonable strategy.  Or, we could avoid marking packets as "lost" on PTO by filtering out those that are already declared lost by ordinary means.